### ISSUE TITLE
Result Problem

### ISSUE THREAD

**Issue by PikaQiu520521** [arneschneuing/DiffSBDD#19]
Hello, thank you for your previous question. Unfortunately, I was not able to fully replicate the process, but I used the same calculation method to evaluate the ligands generated by the model. I have two more questions now. 1. The generated ligand files have "row" and "processed" versions, as well as another folder containing paths and scores. I am not sure what they specifically mean. Could you provide some explanation? 2. The generated ligands do not have the bond relationships between atoms, so they are actually discrete points. How do you think about this issue?

**Comment by PikaQiu520521** [comment]
Hi, thank you very much for your reply. The first reply answered my question very well, but for the second question, I still want to know more about it, I wonder if it is convenient. Observing the files generated by raw and those generated by processed, it can be found that processed only uses the first line of data of each processing result in the raw folder. I haven't seen the processing process yet, and I am confused about it. When visualizing the data, I found that these points are very discrete. Using openbabel for post-processing, I can’t generate new molecules. I don’t know where the problem is, can you answer it?

**Comment by pearl-rabbit** [comment]
Hello, what was the result of the 'Diversity' indicator when evaluating it?
Whether reading molecules from "raw" or "processed", my result is 0, and the value of SA is much smaller than the results in the paper.

The following are the results of evaluating the molecules in raw:
```
QED: 0.495 \pm 0.08
SA: 0.234 \pm 0.03
LogP: 0.021 \pm 1.10
Lipinski: 4.883 \pm 0.34
Diversity: 0.000 \pm 0.00

```

**Comment by pearl-rabbit** [comment]
Hi @arneschneuing , thank you for your timely response. I set n_samples=100 and batch_ size=60 in test.py. If I want to generate multiple molecules for a protein pocket, should I set the 'n_samples' value higher?
 I use the retrained all atom model, and except for the cutoff value set to 1, all other parameters have not been changed.

**Comment by arneschneuing** [comment]
Hi, `n_samples=100` should be fine. It means that 100 valid molecules are generated *per pocket*. How did you provide those molecules to the `evaluate()` function? Did you create a nested list?

A cutoff value of 1 [Å] seems very low. It is less than a typical bond length. Maybe you should consider a higher threshold to create sufficient edges.

**Comment by arneschneuing** [comment]
@PikaQiu520521, I'm sorry but I don't really understand the question.. What do you mean by 'discrete points'? Could you provide an example?

**Comment by pearl-rabbit** [comment]
This is the code for reading molecules:

`from analysis.metrics import MoleculeProperties
from rdkit import Chem
import os

filePath = 'outdir/raw'
filenames = os.listdir(filePath)
pocket_list = []
for filename in filenames:
    suppl = Chem.SDMolSupplier(filePath + filename, sanitize=False)
    mols = [mol for mol in suppl if mol]
    pocket_list.append(mols)
mol_metrics = MoleculeProperties()
all_qed, all_sa, all_logp, all_lipinski, per_pocket_diversity = mol_metrics.evaluate(pocket_list)`

'filePath' is the path of the test set sampling results, storing 100 sdf files, each containing the coordinates and atomic types of   molecules. (Due to some issues with the server where I store the files, I am unable to provide an example of sdf here temporarily.)

**Comment by pearl-rabbit** [comment]
Add some information about SDF (partial)
```
/tmp/tmp0uy5pvze
     RDKit          3D

 10  0  0  0  0  0  0  0  0  0999 V2000
  212.8940  256.9690 -731.9470 C   0  0  0  0  0 15  0  0  0  0  0  0
  102.1060  689.9250  835.7430 C   0  0  0  0  0 15  0  0  0  0  0  0
  373.3720  488.1380   99.3380 N   0  0  0  0  0 15  0  0  0  0  0  0
 -121.1140 -202.4730 -229.0440 O   0  0  0  0  0 15  0  0  0  0  0  0
   81.5140 -312.8040 -889.3270 C   0  0  0  0  0 15  0  0  0  0  0  0
  469.9780  378.1600  -56.7650 C   0  0  0  0  0 15  0  0  0  0  0  0
-1134.4260  -31.9830 -325.5930 C   0  0  0  0  0 15  0  0  0  0  0  0
   90.7670 -750.9430  359.6320 C   0  0  0  0  0 15  0  0  0  0  0  0
  151.3290 -384.8160  940.0640 N   0  0  0  0  0 15  0  0  0  0  0  0
 -455.7040  146.2210  -40.3020 F   0  0  0  0  0 15  0  0  0  0  0  0
M  END
$$$$
/tmp/tmpy5abngjg
     RDKit          3D

 10  0  0  0  0  0  0  0  0  0999 V2000
  496.7010 -652.7640 -387.7500 C   0  0  0  0  0 15  0  0  0  0  0  0
  310.3280  123.9180  362.8120 O   0  0  0  0  0 15  0  0  0  0  0  0
  208.2730  271.3880  292.1010 C   0  0  0  0  0 15  0  0  0  0  0  0
  210.0160  112.3630 -123.2720 C   0  0  0  0  0 15  0  0  0  0  0  0
 -352.2210  599.3190  444.5210 C   0  0  0  0  0 15  0  0  0  0  0  0
 -236.8300   92.5520  125.7750 C   0  0  0  0  0 15  0  0  0  0  0  0
 -116.2680  207.4840  857.0740 C   0  0  0  0  0 15  0  0  0  0  0  0
    7.5230 -224.9410 -509.7420 C   0  0  0  0  0 15  0  0  0  0  0  0
 -659.0520 -704.5070 -360.1970 C   0  0  0  0  0 15  0  0  0  0  0  0
 -105.4780  434.2140 -749.4830 O   0  0  0  0  0 15  0  0  0  0  0  0
M  END
$$$$
/tmp/tmp8sof9qn7
     RDKit          3D

 10  0  0  0  0  0  0  0  0  0999 V2000
  524.6150   75.3350  -19.6130 C   0  0  0  0  0 15  0  0  0  0  0  0
  409.8610 -390.1750   -2.5490 N   0  0  0  0  0 15  0  0  0  0  0  0
  320.3360  -96.5080 -245.8160 C   0  0  0  0  0 15  0  0  0  0  0  0
  682.2410  469.3580  436.2820 C   0  0  0  0  0 15  0  0  0  0  0  0
 -549.7140  251.2520 -245.5920 O   0  0  0  0  0 15  0  0  0  0  0  0
 -904.0780 -441.9380  -15.1740 C   0  0  0  0  0 15  0  0  0  0  0  0
 -467.1750 -397.4860 -607.6640 C   0  0  0  0  0 15  0  0  0  0  0  0
  -65.1000  310.1250 -461.5020 C   0  0  0  0  0 15  0  0  0  0  0  0
  266.0340  333.3790  240.5050 N   0  0  0  0  0 15  0  0  0  0  0  0
 -448.1230  137.2600  860.2680 C   0  0  0  0  0 15  0  0  0  0  0  0
M  END
$$$$
```

**Comment by arneschneuing** [comment]
A cutoff of 3.0 still seems rather low but it might work because information is also propagated through several layers of message passing. However, I haven't tried this value myself yet and can therefore not say if the scores will be similar to the ones from the paper.

Your code for reading molecules looks fine to me. Could you please check how many items your `pocket_list` and each list within `pocket_list` contain before you pass them to `evaluate()`?
```python
print(len(pocket_list))
print([len(x) for x in pocket_list])
```

**Comment by pearl-rabbit** [comment]
If there is no problem,the output is:
```
100
[120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120]
```

That may be due to the value of 'cutoff'.

**Comment by arneschneuing** [comment]
It's hard to tell. The cutoff value could cause your molecules to be less realistic but I'm still surprised by the `0.0` diversity value. Have you visually inspected some of the generated molecules? Do they look more or less like molecules (several atoms connected by bonds) or is there some obvious failure mode (e.g. it always outputs disconnected point clouds)?

**Comment by pearl-rabbit** [comment]
I didn't carefully analyze the reason, just replaced the original method of calculating diversity with the code in the original annotation section, and obtained a non zero value.
[https://github.com/arneschneuing/DiffSBDD/blob/main/analysis/metrics.py#L196](url)
I checked the sdf file and found that the generated molecules have no edges.
AndI retrained the model and obtained effective results evaluating with original code. 

By the way, in the text, regarding ca_only model doesn't have a cutoff set, but it's too big for me , I set a limit on the cutoff and achieved effective results. Does this value only affect the number of edges when generating the graph?

**Comment by arneschneuing** [comment]
Yes, it determines which atoms (nodes) are connected in the graph that the neural network processes.
